src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Navigation.java

```diff
@@ -350,293 +336,58 @@ private void addMainNavLinks(Content tree) {
-                            ANNOTATION_TYPE_MEMBER_OPTIONAL, listContents);
-                } else {
-                    TypeElement typeElement = (TypeElement) element;
-                    for (VisibleMemberTable.Kind kind : summarySet) {
-                        if (kind == ENUM_CONSTANTS && !configuration.utils.isEnum(typeElement)) {
-                            continue;
-                        }
-                        if (kind == CONSTRUCTORS && configuration.utils.isEnum(typeElement)) {
-                            continue;
-                        }
-                        AbstractMemberWriter writer
-                                = ((AbstractMemberWriter) memberSummaryBuilder.getMemberSummaryWriter(kind));
-                        if (writer == null) {
-                            addContentToList(listContents, contents.getNavLinkLabelContent(kind));
-                        } else {
-                            addTypeSummaryLink(memberSummaryBuilder.members(kind),
-                                    memberSummaryBuilder.getVisibleMemberTable(),
-                                    kind, listContents);
-                        }
-                    }
-                }
-                if (!listContents.isEmpty()) {
-                    Content li = HtmlTree.LI(contents.summaryLabel);
-                    li.add(Entity.NO_BREAK_SPACE);
-                    tree.add(li);
-                    addListToNav(listContents, tree);
-                }
-                break;
-
             case MODULE:
-                if (displaySummaryModuleDescLink) {
-                    addContentToList(listContents,
-                            links.createLink(HtmlIds.MODULE_DESCRIPTION, contents.navModuleDescription));
-                } else {
-                    addContentToList(listContents, contents.navModuleDescription);
-                }
-                if (displaySummaryModulesLink) {
-                    addContentToList(listContents,
-                            links.createLink(HtmlIds.MODULES, contents.navModules));
-                } else {
-                    addContentToList(listContents, contents.navModules);
-                }
-                if (displaySummaryPackagesLink) {
-                    addContentToList(listContents,
-                            links.createLink(HtmlIds.PACKAGES, contents.navPackages));
-                } else {
-                    addContentToList(listContents, contents.navPackages);
-                }
-                if (displaySummaryServicesLink) {
-                    addContentToList(listContents,
-                            links.createLink(HtmlIds.SERVICES, contents.navServices));
-                } else {
-                    addContentToList(listContents, contents.navServices);
-                }
+            case PACKAGE:
+            case CLASS:
+            case HELP:
+                List<Content> listContents = subNavLinks.getSubNavLinks();
```

Comment:
I think this looks like where you could stream the list, wrap the entries in `LI` and convert back to a list.